sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,314 @@
+        with tqdm_joblib(
+            tqdm(
+                xs_batches,
+                disable=not show_progress_bar,
+                desc=f"Performing {num_sim_samples} posterior runs in"
+                f"{len(xs_batches)} batches.",
+                total=len(xs_batches),
+            )
+        ) as _:
+            samples: Tensor
+            samples = Parallel(n_jobs=num_workers)(  # pyright: ignore[reportAssignmentType]
+                delayed(infer_posterior_on_batch)(
+                    xs_batch, posterior, num_posterior_samples
+                )
+                for xs_batch in xs_batches
+            )
+    else:
+        pbar = tqdm(
+            total=num_sim_samples,
+            disable=not show_progress_bar,
+            desc=f"Running {num_sim_samples} samples for tarp analysis.",
+        )
+
+        with pbar:
+            samples = []
+            for xs_batch in xs_batches:
+                samples.append(
+                    infer_posterior_on_batch(xs_batch, posterior, num_posterior_samples)
+                )
+                pbar.update(infer_batch_size)
+            samples = torch.cat(samples, dim=1)
+
+    return samples
+
+
+def _check_references(references: Tensor, theta: Tensor) -> Tensor:
+    """construct references"""
+
+    num_dims = theta.shape[-1]
+    num_sims = theta.shape[0]
+
+    if not isinstance(references, Tensor):
+        # obtain min/max per dimension of theta
+        lo = (
+            torch.min(theta, dim=-2).values.min(axis=0).values
+        )  # should be 0 if normalized
+        hi = (
+            torch.max(theta, dim=-2).values.max(axis=0).values
+        )  # should be 1 if normalized
+
+        refpdf = torch.distributions.Uniform(low=lo, high=hi)
+        # sample one reference point for each entry in theta
+        references = refpdf.sample((1, num_sims))
+    else:
+        if len(references.shape) == 2:
+            # add singleton dimension in front
+            references = references.unsqueeze(0)
+
+        if len(references.shape) == 3 and references.shape[0] != 1:
+            raise ValueError(
+                f"""references must be a 2D array with a singular first
+                    dimension, received {references.shape}"""
+            )
+
+        if references.shape[-2] != num_sims:
+            raise ValueError(
+                f"references must have the same number samples as samples,"
+                f"received {references.shape[-2]} != {num_sims}"
+            )
+
+        if references.shape[-1] != num_dims:
+            raise ValueError(
+                "references must have the same number of dimensions as "
+                f"samples or theta, received {references.shape[-1]}"
+                f"!= {num_dims}"
+            )
+
+    return references
+
+
+def run_tarp(
+    samples: Tensor,
+    theta: Tensor,
+    references: Optional[Tensor] = None,
+    distance: Callable = l2,
+    num_bins: Optional[int] = None,
+    do_norm: bool = False,
```

Comment:
I suggest to refactor this function quite a bit to match the SBC logic. My idea that it gets as input
```suggestion
    thetas: Tensor (samples from the prior)
    xs: Tensor (corresponding simulations)
    posterior: NeuralPosterior
    num_posterior_samples: int = 1000
    reference_dist: Optional[Distribution] | None = None
    distance: Optional[Callable] = l2, 
    num_workers: int = 1
    show_progress_bar: bool =True
```

The workflow of the function would be

1. check the shapes
2. check whether reference dist is provided, if not, construct from min max values
3. then, have one loop (that will be parallelized with `num_workers` in another PR) that does: 
- for each theta, x in `zip(thetas, xs)`: 
    - sample `num_posterior_samples` from posterior given x
    - sample a reference
    - for each posterior sample, calculate distance to reference and to theta
    - average the count where d-ref < d-theta
    - collect the f_i value

4. return the f_i values (I suggest to do the binning etc outside this function because user might want to use their custom binning or statistics).

Reply:
I hope I did this with 33ca7efa